File: internal/projects/projectconfig.go
package projects

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// ProjectConfigDirName is the per-repo directory for TopoDuctor data (at git worktree root).
const ProjectConfigDirName = ".topoductor"

// ProjectConfigFileName is the JSON file inside ProjectConfigDirName.
const ProjectConfigFileName = "project.json"

// LegacyProjectConfigFileName was the previous single-file location at repo root.
const LegacyProjectConfigFileName = "topoductor.project.json"

// ProjectConfigFile is the on-disk shape of project.json under .topoductor/.
type ProjectConfigFile struct {
	Scripts ProjectScripts `json:"scripts"`
}

// ProjectScripts are shell commands run in a worktree directory (optional each).
type ProjectScripts struct {
	Setup   string `json:"setup,omitempty"`
	Run     string `json:"run,omitempty"`
	Archive string `json:"archive,omitempty"`
}

func projectConfigPath(repoRoot string) string {
	return filepath.Join(filepath.Clean(repoRoot), ProjectConfigDirName, ProjectConfigFileName)
}

func legacyProjectConfigPath(repoRoot string) string {
	return filepath.Join(filepath.Clean(repoRoot), LegacyProjectConfigFileName)
}

func readProjectConfigFile(path string) (ProjectScripts, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return ProjectScripts{}, err
	}
	var f ProjectConfigFile
	if err := json.Unmarshal(data, &f); err != nil {
		return ProjectScripts{}, err
	}
	return f.Scripts, nil
}

// ReadProjectConfig loads scripts from repoRoot/.topoductor/project.json.
// If missing, reads legacy repoRoot/topoductor.project.json when present.
// No file returns empty scripts and nil error.
func ReadProjectConfig(repoRoot string) (ProjectScripts, error) {
	p := projectConfigPath(repoRoot)
	s, err := readProjectConfigFile(p)
	if err == nil {
		return s, nil
	}
	if !os.IsNotExist(err) {
		return ProjectScripts{}, err
	}
	legacy := legacyProjectConfigPath(repoRoot)
	s, err = readProjectConfigFile(legacy)
	if err != nil {
		if os.IsNotExist(err) {
			return ProjectScripts{}, nil
		}
		return ProjectScripts{}, err
	}
	return s, nil
}

// SaveProjectScripts writes repoRoot/.topoductor/project.json (creates the directory).
// If all script fields are empty after trim, removes that file and prunes .topoductor when empty;
// also removes legacy topoductor.project.json if present.
func SaveProjectScripts(repoRoot string, s ProjectScripts) error {
	root := filepath.Clean(repoRoot)
	s.Setup = strings.TrimSpace(s.Setup)
	s.Run = strings.TrimSpace(s.Run)
	s.Archive = strings.TrimSpace(s.Archive)
	p := projectConfigPath(repoRoot)
	dir := filepath.Join(root, ProjectConfigDirName)
	_ = os.Remove(legacyProjectConfigPath(root))
	if s.Setup == "" && s.Run == "" && s.Archive == "" {
		if err := os.Remove(p); err != nil && !os.IsNotExist(err) {
			return err
		}
		_ = os.Remove(dir)
		return nil
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(ProjectConfigFile{Scripts: s}, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(p, data, 0o644)
}

// ExpandScriptPlaceholders replaces {path} with a shell-quoted absolute path (same convention as exit custom cmd).
func ExpandScriptPlaceholders(script, worktreePath string) string {
	return strings.ReplaceAll(script, "{path}", strconv.Quote(worktreePath))
}

File: src/projects/projectConfig.ts
import fs from "node:fs/promises";
import path from "node:path";

export const ProjectConfigDirName = ".topoductor";
export const ProjectConfigFileName = "project.json";
const LegacyProjectConfigFileName = "topoductor.project.json";

export type ProjectScripts = {
  setup?: string;
  run?: string;
  archive?: string;
};

type ProjectConfigFile = {
  scripts: ProjectScripts;
};

function projectConfigPath(repoRoot: string): string {
  return path.join(path.normalize(repoRoot), ProjectConfigDirName, ProjectConfigFileName);
}

function legacyProjectConfigPath(repoRoot: string): string {
  return path.join(path.normalize(repoRoot), LegacyProjectConfigFileName);
}

async function readScriptsFrom(p: string): Promise<ProjectScripts> {
  const data = await fs.readFile(p, "utf8");
  const f = JSON.parse(data) as ProjectConfigFile;
  return f.scripts ?? {};
}

export async function readProjectConfig(
  repoRoot: string
): Promise<ProjectScripts> {
  const p = projectConfigPath(repoRoot);
  try {
    return await readScriptsFrom(p);
  } catch (e) {
    const err = e as NodeJS.ErrnoException;
    if (err.code !== "ENOENT") {
      throw e;
    }
  }
  const legacy = legacyProjectConfigPath(repoRoot);
  try {
    return await readScriptsFrom(legacy);
  } catch (e) {
    const err = e as NodeJS.ErrnoException;
    if (err.code === "ENOENT") {
      return {};
    }
    throw e;
  }
}

export async function saveProjectScripts(
  repoRoot: string,
  s: ProjectScripts
): Promise<void> {
  const root = path.normalize(repoRoot);
  let setup = (s.setup ?? "").trim();
  let run = (s.run ?? "").trim();
  let archive = (s.archive ?? "").trim();
  const p = projectConfigPath(root);
  const dir = path.join(root, ProjectConfigDirName);
  try {
    await fs.unlink(legacyProjectConfigPath(root));
  } catch {
    /* ignore */
  }
  if (!setup && !run && !archive) {
    try {
      await fs.unlink(p);
    } catch (e) {
      const err = e as NodeJS.ErrnoException;
      if (err.code !== "ENOENT") {
        throw e;
      }
    }
    try {
      await fs.rmdir(dir);
    } catch {
      /* ignore */
    }
    return;
  }
  await fs.mkdir(dir, { recursive: true, mode: 0o755 });
  const data = JSON.stringify(
    { scripts: { setup, run, archive } } satisfies ProjectConfigFile,
    null,
    2
  );
  await fs.writeFile(p, data, { mode: 0o644 });
}

export function expandScriptPlaceholders(
  script: string,
  worktreePath: string
): string {
  const q = JSON.stringify(worktreePath);
  return script.replaceAll("{path}", q);
}
